### PR TITLE
[enhancement](Nereids) planner performance speed up

### PR DESCRIPTION
**Pull request by sohardforaname**
# Proposed changes

Issue Number: close #xxx

## Problem summary

optimize planner by:
- reduce duplicated calculation on equals, getOutput, computeOutput eq.
- getOnClauseUsedSlots: the two side of equalTo is centainly slot, so not need to use List.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostEstimate.java [hunk 1]
```diff
@@ -90,11 +90,27 @@ public static CostEstimate ofMemory(double memoryCost) {
     /**
      * Sums partial cost estimates of some (single) plan node.
      */
+    @Deprecated
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostEstimate.java:93]: No rename it, just remove it.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostEstimate.java [hunk 2]
```diff
@@ -88,13 +86,17 @@ public static CostEstimate ofMemory(double memoryCost) {
...
      */
     public static CostEstimate sum(CostEstimate one, CostEstimate two, CostEstimate... more) {
-        return Stream.concat(Stream.of(one, two), Stream.of(more))
-                .reduce(zero(), (a, b) -> new CostEstimate(
-                        a.cpuCost + b.cpuCost,
-                        a.memoryCost + b.memoryCost,
-                        a.networkCost + b.networkCost));
+        double v1 = one.cpuCost + two.cpuCost;
+        double v2 = one.memoryCost + two.memoryCost;
+        double v3 = one.networkCost + one.networkCost;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostEstimate.java:94]: use meaningful variable names

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecHash.java [hunk 3]
```diff
@@ -208,16 +214,18 @@ public boolean equals(Object o) {
...
         DistributionSpecHash that = (DistributionSpecHash) o;
-        return tableId == that.tableId && orderedShuffledColumns.equals(that.orderedShuffledColumns)
-                && shuffleType == that.shuffleType && partitionIds.equals(that.partitionIds)
-                && equivalenceExprIds.equals(that.equivalenceExprIds)
-                && exprIdToEquivalenceSet.equals(that.exprIdToEquivalenceSet);
+        return shuffleType == that.shuffleType && orderedShuffledColumns.equals(that.orderedShuffledColumns);
+        // return tableId == that.tableId && orderedShuffledColumns.equals(that.orderedShuffledColumns)
+        //         && shuffleType == that.shuffleType && partitionIds.equals(that.partitionIds)
+        //         && equivalenceExprIds.equals(that.equivalenceExprIds)
+        //         && exprIdToEquivalenceSet.equals(that.exprIdToEquivalenceSet);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecHash.java:221]: remove it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecHash.java [hunk 4]
```diff
@@ -208,16 +214,18 @@ public boolean equals(Object o) {
...
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(orderedShuffledColumns, shuffleType, tableId, partitionIds,
-                equivalenceExprIds, exprIdToEquivalenceSet);
+        // return Objects.hash(orderedShuffledColumns, shuffleType, tableId, partitionIds,
+        //         equivalenceExprIds, exprIdToEquivalenceSet);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecHash.java:227]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java [hunk 5]
```diff
@@ -69,11 +69,11 @@ public List<List<PhysicalProperties>> getRequestChildrenPropertyList(GroupExpres
 
     @Override
     public Void visit(Plan plan, PlanContext context) {
-        List<PhysicalProperties> requiredPropertyList = Lists.newArrayList();
-        for (int i = 0; i < context.getGroupExpression().arity(); i++) {
-            requiredPropertyList.add(PhysicalProperties.ANY);
+        List<PhysicalProperties> nCopyList = Lists.newArrayListWithCapacity(context.getGroupExpression().arity());
+        for (int i = context.getGroupExpression().arity(); i > 0; --i) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java:73]: why use -- ?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 6]
```diff
@@ -68,16 +68,17 @@ private static class JoinSlotCoverageChecker {
...
+        JoinSlotCoverageChecker(Set<ExprId> left, Set<ExprId> right) {
+            leftExprIds = left;
+            rightExprIds = right;
         }
 
-        boolean isCoveredByRightSlots(Set<Slot> slots) {
-            return slots.stream()
-                    .map(Slot::getExprId)
-                    .allMatch(rightExprIds::contains);
+        boolean isCoveredByLeftSlots(ExprId slot) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:76]: add comment to explain why only one slot is enough

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 7]
```diff
@@ -152,37 +154,35 @@ public static Pair<List<Expression>, List<Expression>> extractExpressionForHashT
      */
     public static Pair<List<ExprId>, List<ExprId>> getOnClauseUsedSlots(
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:155]: make sure these functions has enough unit test.

**Review by morrySnow — APPROVED** [review]

**Pull request merged**